File: ios/Sources/WanxiangBook/BookSource/Legado/LegadoRuleParser.swift
//
//  LegadoRuleParser.swift
//  万象书屋 iOS · legado 规则字符串切分
//
//  对应 Android RuleAnalyzer.splitRule + AnalyzeRule.splitSourceRule
//
//  关键: `&&` `||` `%%` 在 [...] 和 (...) 平衡组里要忽略, 否则
//   ".item:nth-child(2)" 的 ":2" 会被当成正则.
//

import Foundation

public enum LegadoRuleParser {

    /// 顶层入口: 一段 ruleStr → [SourceRule]
    /// 拆 && / || / %%, 但 [...] (...) 内的不拆.
    /// 切完后每段:
    ///   - 推断 mode (prefix / format hint)
    ///   - 切 ##regex##replace[##]
    ///   - 标占位符 (Mode 升级)
    public static func split(_ ruleStr: String) -> [LegadoSourceRule] {
        guard !ruleStr.isEmpty else { return [] }
        let chunks = splitTop(ruleStr, separators: ["&&", "||"])
        return chunks.map { parseSingle(String($0)) }
    }

    // MARK: - 单段解析

    static func parseSingle(_ s: String) -> LegadoSourceRule {
        var raw = s.trimmingCharacters(in: .whitespacesAndNewlines)
        // 万象书屋: 处理 "@@" 强制 Default
        if raw.hasPrefix("@@") {
            raw = String(raw.dropFirst(2))
        }
        // 显式前缀
        var mode: LegadoMode = .css
        var isAllInOneRegex = false
        let lower = raw.lowercased()
        if lower.hasPrefix("@css:") {
            mode = .css; raw = String(raw.dropFirst(5))
        } else if lower.hasPrefix("@xpath:") {
            mode = .xpath; raw = String(raw.dropFirst(7))
        } else if lower.hasPrefix("@json:") {
            mode = .json; raw = String(raw.dropFirst(6))
        } else if lower.hasPrefix("@js:") {
            mode = .js; raw = String(raw.dropFirst(4))
        } else if lower.hasPrefix("@regex:") {
            mode = .regex; raw = String(raw.dropFirst(7))
        } else if raw.hasPrefix(":") {
            // 万象书屋: 正则之 AllInOne, 只能在 searchList / exploreList / bookInfoInit / tocList 等列表场景使用
            // 语法 `:pattern`, 后续字段可用 `$1` `$2` 抽捕获组
            mode = .regex; raw = String(raw.dropFirst()); isAllInOneRegex = true
        } else if raw.hasPrefix("$.") || raw.hasPrefix("$[") {
            mode = .json
        } else if raw.hasPrefix("//") {
            mode = .xpath
        } else if raw.hasPrefix("/") && !raw.hasPrefix("//") {
            // 万象书屋: legado 单 / 也是 XPath (//html 的简化), 但 "/path" URL 不是
            // 经验: 含 [@ 或 @text 的 / 开头 = XPath
            if raw.contains("[@") || raw.contains("/text(") || raw.contains("/@") {
                mode = .xpath
            }
        }
        // {{...}} / @get: / <js>...</js> 占位符 → 升级 Regex 模板模式
        let hasPlaceholder = raw.contains("{{") || raw.contains("@get:") || raw.contains("<js>")

        // 切 ##regex##replace[##]
        var rule = raw
        var replaceRegex = ""
        var replacement = ""
        var replaceFirst = false
        if rule.contains("##") {
            let parts = splitChainSafe(rule, separator: "##")
            rule = String(parts[0])
            if parts.count >= 2 { replaceRegex = String(parts[1]) }
            if parts.count >= 3 { replacement = String(parts[2]) }
            if parts.count >= 4 { replaceFirst = true }
        }

        // 占位符模式且没有 select 类前缀, mode 升级 regex
        if hasPlaceholder, mode == .css {
            if rule.hasPrefix("{{") && rule.hasSuffix("}}") {
                mode = .regex
            } else if rule.hasPrefix("<js>") {
                // 万象书屋: legado bookList 经常整段就是 `<js>(function(){... return arr;})()</js>`
                // 此时 JS 直接产出列表 (而非"求值后字符串再选" 模板模式)
                // 把它当 .js 走 runJS, 返回数组时按元素散开成 list of JSON strings
                let trimmed = rule.trimmingCharacters(in: .whitespacesAndNewlines)
                if trimmed.hasPrefix("<js>"), let end = trimmed.range(of: "</js>"),
                   trimmed[end.upperBound...].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    mode = .js
                    rule = String(trimmed[trimmed.index(trimmed.startIndex, offsetBy: 4)..<end.lowerBound])
                } else {
                    mode = .regex
                }
            } else if rule.hasPrefix("http") || rule.hasPrefix("/") {
                // URL 模板
                mode = .regex
            }
        }

        return LegadoSourceRule(
            mode: mode,
            rule: rule.trimmingCharacters(in: .whitespacesAndNewlines),
            replaceRegex: replaceRegex,
            replacement: replacement,
            replaceFirst: replaceFirst,
            isAllInOneRegex: isAllInOneRegex,
            hasPlaceholder: hasPlaceholder
        )
    }

    // MARK: - 平衡组感知的分割

    /// 把字符串按 separators 分割, 但 [...] (...) <...> 平衡组里的不分.
    /// 字符串 "abc&&[def&&ghi]&&jkl" 分成 ["abc", "[def&&ghi]", "jkl"]
    static func splitTop(_ s: String, separators: [String]) -> [Substring] {
        var out: [Substring] = []
        var depthSquare = 0   // [ ]
        var depthRound = 0    // ( )
        var depthAngle = 0    // < > (用于 <js>)
        var inSingleQ = false
        var inDoubleQ = false

        let chars = Array(s)
        var sliceStart = 0
        var i = 0
        while i < chars.count {
            let c = chars[i]
            // 引号 (legado 平衡组规则: 单/双引号内不算嵌套)
            if !inDoubleQ, c == "'" { inSingleQ.toggle(); i += 1; continue }
            if !inSingleQ, c == "\"" { inDoubleQ.toggle(); i += 1; continue }
            if inSingleQ || inDoubleQ { i += 1; continue }

            switch c {
            case "[": depthSquare += 1
            case "]": depthSquare = max(0, depthSquare - 1)
            case "(": depthRound += 1
            case ")": depthRound = max(0, depthRound - 1)
            case "<":
                // 只对 <js> 等明显的 tag 计深度, 避免与 < 比较运算符冲突
                let rest = chars[i...]
                if rest.starts(with: "<js>") || rest.starts(with: "</js>") {
                    if rest.starts(with: "</js>") {
                        depthAngle = max(0, depthAngle - 1)
                        i += 5; continue
                    } else {
                        depthAngle += 1
                        i += 4; continue
                    }
                }
            default: break
            }

            // 仅顶层 (所有 depth 都 0) 才 try 切
            var matchedSep = false
            if depthSquare == 0, depthRound == 0, depthAngle == 0 {
                for sep in separators {
                    let sepChars = Array(sep)
                    if i + sepChars.count <= chars.count,
                       Array(chars[i..<(i + sepChars.count)]) == sepChars {
                        let segment = String(chars[sliceStart..<i])
                        out.append(Substring(segment))
                        sliceStart = i + sepChars.count
                        i += sepChars.count
                        matchedSep = true
                        break
                    }
                }
            }
            if !matchedSep { i += 1 }
        }
        // 最后一段
        let last = String(chars[sliceStart..<chars.count])
        if !last.isEmpty || !out.isEmpty {
            out.append(Substring(last))
        }
        return out
    }

    /// 安全切 ##: 不切 [...] (...) 内的 ##
    static func splitChainSafe(_ s: String, separator: String) -> [Substring] {
        return splitTop(s, separators: [separator])
    }
}

File: rn/src/engine/LegadoRuleParser.ts
/**
 * 万象书屋 RN · legado 规则字符串切分
 * 对应 iOS: LegadoRuleParser.swift
 * 对应 Android: RuleAnalyzer.splitRule + AnalyzeRule.splitSourceRule
 *
 * 关键: `&&` `||` `%%` 在 [...] 和 (...) 平衡组里要忽略
 */

import {LegadoMode, LegadoSourceRule} from './types';

/**
 * 顶层入口: 一段 ruleStr → [LegadoSourceRule]
 */
export function splitRules(ruleStr: string): LegadoSourceRule[] {
  if (!ruleStr) return [];
  const chunks = splitTop(ruleStr, ['&&', '||']);
  return chunks.map(parseSingle);
}

/**
 * 解析单段规则 (mode 推断 + ## 切分)
 */
export function parseSingle(s: string): LegadoSourceRule {
  let raw = s.trim();

  // @@ 强制默认 CSS
  if (raw.startsWith('@@')) {
    raw = raw.slice(2);
  }

  // 显式前缀
  let mode: LegadoMode = LegadoMode.CSS;
  let isAllInOneRegex = false;
  const lower = raw.toLowerCase();

  if (lower.startsWith('@css:')) {
    mode = LegadoMode.CSS;
    raw = raw.slice(5);
  } else if (lower.startsWith('@xpath:')) {
    mode = LegadoMode.XPath;
    raw = raw.slice(7);
  } else if (lower.startsWith('@json:')) {
    mode = LegadoMode.JSON;
    raw = raw.slice(6);
  } else if (lower.startsWith('@js:')) {
    mode = LegadoMode.JS;
    raw = raw.slice(4);
  } else if (lower.startsWith('@regex:')) {
    mode = LegadoMode.Regex;
    raw = raw.slice(7);
  } else if (raw.startsWith(':')) {
    // AllInOne regex (用于列表场景)
    mode = LegadoMode.Regex;
    raw = raw.slice(1);
    isAllInOneRegex = true;
  } else if (raw.startsWith('$.') || raw.startsWith('$[')) {
    mode = LegadoMode.JSON;
  } else if (raw.startsWith('//')) {
    mode = LegadoMode.XPath;
  } else if (raw.startsWith('/')) {
    // 单 / 含 XPath 特征的判为 XPath
    if (raw.includes('[@') || raw.includes('/text(') || raw.includes('/@')) {
      mode = LegadoMode.XPath;
    }
  }

  // 占位符检测
  const hasPlaceholder =
    raw.includes('{{') || raw.includes('@get:') || raw.includes('<js>');

  // 切 ##regex##replace[##]
  let rule = raw;
  let replaceRegex = '';
  let replacement = '';
  let replaceFirst = false;

  if (rule.includes('##')) {
    const parts = splitChainSafe(rule, '##');
    rule = parts[0];
    if (parts.length >= 2) replaceRegex = parts[1];
    if (parts.length >= 3) replacement = parts[2];
    if (parts.length >= 4) replaceFirst = true;
  }

  // 占位符模式升级
  if (hasPlaceholder && mode === LegadoMode.CSS) {
    if (rule.startsWith('{{') && rule.endsWith('}}')) {
      mode = LegadoMode.Regex;
    } else if (rule.startsWith('<js>')) {
      // 整段 <js>...</js> → JS 模式
      const trimmed = rule.trim();
      const endIdx = trimmed.indexOf('</js>');
      if (
        trimmed.startsWith('<js>') &&
        endIdx > 0 &&
        trimmed.slice(endIdx + 5).trim() === ''
      ) {
        mode = LegadoMode.JS;
        rule = trimmed.slice(4, endIdx);
      } else {
        mode = LegadoMode.Regex;
      }
    } else if (rule.startsWith('http') || rule.startsWith('/')) {
      mode = LegadoMode.Regex;
    } else {
      mode = LegadoMode.Regex;
    }
  }

  return {
    mode,
    rule: rule.trim(),
    replaceRegex,
    replacement,
    replaceFirst,
    isAllInOneRegex,
    hasPlaceholder,
  };
}

/**
 * 平衡组感知的分割
 * 字符串按 separators 切分，但 [...] (...) <js>...</js> 引号内不切
 *
 * implicitJsChainSplit: 段中间出现的 @js: 是否当隐式 && 切
 */
export function splitTop(
  s: string,
  separators: string[],
  implicitJsChainSplit: boolean = true,
): string[] {
  const out: string[] = [];
  let depthSquare = 0;
  let depthRound = 0;
  let depthAngle = 0;
  let inSingleQ = false;
  let inDoubleQ = false;
  let inJsPrefix = false;

  const chars = s;
  let sliceStart = 0;
  let i = 0;

  while (i < chars.length) {
    const c = chars[i];

    // 引号
    if (!inDoubleQ && c === "'") {
      inSingleQ = !inSingleQ;
      i++;
      continue;
    }
    if (!inSingleQ && c === '"') {
      inDoubleQ = !inDoubleQ;
      i++;
      continue;
    }
    if (inSingleQ || inDoubleQ) {
      i++;
      continue;
    }

    // 嵌套追踪
    if (c === '[') depthSquare++;
    else if (c === ']') depthSquare = Math.max(0, depthSquare - 1);
    else if (c === '(') depthRound++;
    else if (c === ')') depthRound = Math.max(0, depthRound - 1);
    else if (c === '<') {
      const rest = chars.slice(i);
      if (rest.startsWith('</js>')) {
        depthAngle = Math.max(0, depthAngle - 1);
        i += 5;
        continue;
      } else if (rest.startsWith('<js>')) {
        depthAngle++;
        i += 4;
        continue;
      }
    } else if (c === '@') {
      const rest = chars.slice(i);
      if (rest.startsWith('@js:')) {
        if (
          implicitJsChainSplit &&
          depthSquare === 0 &&
          depthRound === 0 &&
          depthAngle === 0 &&
          !inJsPrefix
        ) {
          const pre = chars.slice(sliceStart, i);
          const trimmed = pre.trim();
          if (trimmed.length > 0) {
            out.push(pre);
            sliceStart = i;
            inJsPrefix = true;
            i += 4;
            continue;
          }
        }
        inJsPrefix = true;
        i += 4;
        continue;
      }
    }

    // 尝试分割（仅在顶层且非 @js: 段内）
    let matchedSep = false;
    if (
      depthSquare === 0 &&
      depthRound === 0 &&
      depthAngle === 0 &&
      !inJsPrefix
    ) {
      for (const sep of separators) {
        if (chars.slice(i, i + sep.length) === sep) {
          out.push(chars.slice(sliceStart, i));
          sliceStart = i + sep.length;
          i += sep.length;
          matchedSep = true;
          break;
        }
      }
    }
    if (!matchedSep) i++;
  }

  // 最后一段
  const last = chars.slice(sliceStart);
  if (last.length > 0 || out.length > 0) {
    out.push(last);
  }

  return out;
}

/**
 * 安全切 ##: 不切 [...] (...) 内的 ##
 */
export function splitChainSafe(s: string, separator: string): string[] {
  return splitTop(s, [separator]);
}
